source/RoslynAnalyzers/MessageContractAnalyzers.cs

```diff
@@ -0,0 +1,302 @@
+        // we're being a little bit tricky here. Normally you look up something in CompilationStart, then register actions
+        // which depend on that data. Those child actions then only run inside the compilation execution, which doesn't happen
+        // super often. This leads to errors only updating periodically in the IDE.
+        // We don't actually want to wait for compilation start; we just use it as a convenient way to cache some type lookups
+        // hence disabling this warning. If the cache isn't populated everything still works, just slightly less quickly.
+        void CacheCommonTypes(CompilationStartAnalysisContext context)
+        {
+            cachedTypeStorage = LookupCommonTypes(context.Compilation);
+        }
+#pragma warning restore RS1012
+
+        // ReSharper disable once InconsistentNaming
+        record SpecialTypeDeclarations(
+            INamedTypeSymbol Boolean,
+            INamedTypeSymbol String,
+            INamedTypeSymbol IEnumerable,
+            INamedTypeSymbol? SpaceId,
+            INamedTypeSymbol? CaseInsensitiveStringTinyType);
+
+        // perf opt so we only lookup commonly used types by metadata once
+        SpecialTypeDeclarations LookupCommonTypes(Compilation compilation)
+             => cachedTypeStorage ?? new(
+                Boolean: compilation.GetSpecialType(SpecialType.System_Boolean),
+                String: compilation.GetSpecialType(SpecialType.System_String),
+                IEnumerable: compilation.GetSpecialType(SpecialType.System_Collections_IEnumerable),
+                SpaceId: compilation.GetTypeByMetadataName("Octopus.Server.MessageContracts.Features.Spaces.SpaceId"),
+                CaseInsensitiveStringTinyType: compilation.GetTypeByMetadataName("Octopus.TinyTypes.CaseInsensitiveStringTinyType"));
+
+        static readonly Regex EventNameRegex = new("(?<!V\\d+)Event(V\\d+)?$", RegexOptions.Compiled | RegexOptions.CultureInvariant);
+        static readonly Regex RequestNameRegex = new("(?<!V\\d+)Request(V\\d+)*$", RegexOptions.Compiled | RegexOptions.CultureInvariant);
+        static readonly Regex CommandNameRegex = new("(?<!V\\d+)Command(V\\d+)*$", RegexOptions.Compiled | RegexOptions.CultureInvariant);
+
+        const string TypeNameIRequest = "IRequest";
+        const string TypeNameICommand = "ICommand";
+        const string TypeNameIResponse = "IResponse";
+        const string TypeNameIEvent = "IEvent";
+        const string TypeNameResource = "Resource";
+
+        void CheckNode(SyntaxNodeAnalysisContext context)
+        {
+            if (context.Node is not TypeDeclarationSyntax typeDec) return; // we are only interested in class or struct declarations
+
+            // This analyzer inspects "API Surface" types; i.e. Commands, Requests, Responses, Events and Resources
+            // first we have to figure out which kind of type we've encountered (if any)
+            var apiSurfaceType = DetermineApiSurfaceType(typeDec);
```

Comment:
@matt-richardson I broke the "what kind of thing are we looking at" into it's own function per your earlier comment. Would you mind having another look?

Reply:
Yeah - that looks better. Like the detailed (and useful) comments